File: pkg/rtweekend/bvh.go
// BVH (Bounding Volume Hierarchy) Node
package rtweekend

import (
	"log"
	"sort"
)

type BVH_Node struct {
	left  interface{}
	right interface{}
	box   AABB
}

func NewBVH_NodeFromArr(
	srcObjects []Hittable,
	start, end uint,
	time0, time1 float64,
) *BVH_Node {
	bvhNode := new(BVH_Node)

	var boxLeft, boxRight AABB
	axis := RandomIntInRange(0, 2)
	objectSpan := end - start

	comparator := func(a, b *Hittable) bool {
		return boxCompare(a, b, axis)
	}

	if objectSpan == 1 {
		bvhNode.left = srcObjects[start]
		bvhNode.right = srcObjects[start]
	} else if objectSpan == 2 {
		if comparator(&srcObjects[start], &srcObjects[start]) {
			bvhNode.left = srcObjects[start]
			bvhNode.right = srcObjects[start+1]
		} else {
			bvhNode.left = srcObjects[start+1]
			bvhNode.right = srcObjects[start]
		}
	} else {
		sort.Slice(
			srcObjects,
			func(i, j int) bool {
				return comparator(&srcObjects[i], &srcObjects[j])
			},
		)
		mid := start + objectSpan/2
		bvhNode.left = *NewBVH_NodeFromArr(srcObjects, start, mid, time0, time1)
		bvhNode.right = *NewBVH_NodeFromArr(srcObjects, mid, end, time0, time1)
	}

	if !boxLeft.BoundingBox(time0, time1, &boxLeft) ||
		!boxRight.BoundingBox(time0, time1, &boxRight) {
		log.Println("No bounding box in NewBVH_NodeFromArr")
	}

	bvhNode.box = *SurroundingBox(&boxLeft, &boxRight)

	return bvhNode
}

func NewBVH_Node(list *HittableList, time0, time1 float32) {
}

func boxCompare(a, b *Hittable, axis int) bool {
	var boxA, boxB AABB

	if !(*a).BoundingBox(0, 0, &boxA) || !(*b).BoundingBox(0, 0, &boxB) {
		log.Println("No bounding box in BVH_Node")
	}

	return boxA.Min().E(axis) < boxA.Min().E(axis)
}

func BoxXCompare(a, b *Hittable) bool {
	return boxCompare(a, b, 0)
}

func BoxYCompare(a, b *Hittable) bool {
	return boxCompare(a, b, 1)
}

func BoxZCompare(a, b *Hittable) bool {
	return boxCompare(a, b, 2)
}

File: rtiow/src/hittable/bvh.rs
use super::{Aabb, HitRecord, Hittable, HittableList};
use crate::{interval::Interval, ray::Ray, vec3::Point3};
use anyhow::Result;
use rayon::prelude::*;
use std::{cmp::Ordering, sync::Arc};

#[derive(Debug)]
pub struct BvhNode {
    left: Arc<dyn Hittable>,
    right: Arc<dyn Hittable>,
    bbox: Aabb,
}

impl BvhNode {
    pub fn new(objects: &mut Vec<Arc<dyn Hittable>>, start: usize, end: usize) -> Result<Self> {
        let mut bbox = Aabb::EMPTY;

        for object_index in start..end {
            bbox = Aabb::surrounding_box(
                &bbox,
                &objects[object_index]
                    .bounding_box()
                    .expect("Bounding box not available"),
            );
        }

        let axis = bbox.longest_axis();
        let comparator = match axis {
            0 => Self::box_x_compare,
            1 => Self::box_y_compare,
            _ => Self::box_z_compare,
        };
        let object_span = end - start;

        let (left, right) = match object_span {
            1 => (objects[start].clone(), objects[start].clone()),
            2 => (objects[start].clone(), objects[start + 1].clone()),
            _ => {
                objects[start..end].par_sort_by(|a, b| comparator(a, b));

                let mid = start + object_span / 2;
                let left: Arc<dyn Hittable> = Arc::new(Self::new(objects, start, mid)?);
                let right: Arc<dyn Hittable> = Arc::new(Self::new(objects, mid, end)?);

                (left, right)
            }
        };

        Ok(Self { left, right, bbox })
    }

    pub fn from(list: &HittableList) -> Result<Self> {
        let mut objects = list.objects.clone();
        let length = objects.len();

        Self::new(&mut objects, 0, length)
    }

    fn box_compare(a: &Arc<dyn Hittable>, b: &Arc<dyn Hittable>, axis_index: usize) -> Ordering {
        let a_axis_interval = a
            .bounding_box()
            .unwrap_or(Aabb::new(Point3::zeroes(), Point3::zeroes()));
        let b_axis_interval = b
            .bounding_box()
            .unwrap_or(Aabb::new(Point3::zeroes(), Point3::zeroes()));

        a_axis_interval
            .axis_interval(axis_index)
            .min
            .partial_cmp(&b_axis_interval.axis_interval(axis_index).min)
            .expect("No bounding box in Bvh::new")
    }

    fn box_x_compare(a: &Arc<dyn Hittable>, b: &Arc<dyn Hittable>) -> Ordering {
        Self::box_compare(a, b, 0)
    }

    fn box_y_compare(a: &Arc<dyn Hittable>, b: &Arc<dyn Hittable>) -> Ordering {
        Self::box_compare(a, b, 1)
    }

    fn box_z_compare(a: &Arc<dyn Hittable>, b: &Arc<dyn Hittable>) -> Ordering {
        Self::box_compare(a, b, 2)
    }
}

impl Hittable for BvhNode {
    fn hit(&self, r: &Ray, ray_t: &Interval) -> Option<HitRecord> {
        if !self.bbox.hit(r, ray_t) {
            return None;
        }

        let hit_left = if let Some(rec) = self.left.hit(r, ray_t) {
            Some(rec)
        } else {
            None
        };
        let t = hit_left.as_ref().map(|rec| rec.t).unwrap_or(ray_t.max);
        let hit_right = if let Some(rec) = self.right.hit(r, &Interval::new(ray_t.min, t)) {
            Some(rec)
        } else {
            None
        };

        hit_left.or_else(|| hit_right)
    }

    fn bounding_box(&self) -> Option<Aabb> {
        Some(self.bbox.clone())
    }
}
